onnxruntime/contrib_ops/rocm/math/gemm_float8.cu

```diff
@@ -0,0 +1,210 @@
+    transA_ = info.GetAttrOrDefault<int64_t>("transA", 0);
+    transB_ = info.GetAttrOrDefault<int64_t>("transB", 0);
+    dtype_ = info.GetAttrOrDefault<int64_t>("dtype", onnx::TensorProto_DataType_FLOAT16);
+    alpha_ = info.GetAttrOrDefault<float>("alpha", 1);
+    beta_ = info.GetAttrOrDefault<float>("beta", 0);
+  }
+  Status ComputeInternal(OpKernelContext* ctx) const override;
+
+ private:
+#if !defined(DISABLE_FLOAT8_TYPES)
+  template <typename Fp8T>
+  Status ComputeFp8Fp16Fp16(OpKernelContext* ctx, const Tensor* A, const Tensor* scaleA, const Tensor* B, Tensor* C) const;
+  template <typename Fp8T>
+  Status ComputeFp16Fp8Fp16(OpKernelContext* ctx, const Tensor* A, const Tensor* B, const Tensor* scaleB, Tensor* C) const;
+
+  template <typename TA, typename TB, typename TC, BlasOp OpA, BlasOp OpB>
+  [[nodiscard]] inline auto* GetOp() const {
+    using OpT = GemmFloat8TunableOp<TA, TB, TC, OpA, OpB>;
+    if (tunable_op_) {
+      return static_cast<OpT*>(tunable_op_.get());
+    }
+
+    auto create = std::make_unique<OpT>();   // avoid new
+    tunable_op_ = std::shared_ptr<void>(create.release(), [](void* ptr) {
+      auto release = std::unique_ptr<OpT>();  // avoid delete
+      release.reset(static_cast<OpT*>(ptr));
+    });
+
+    return static_cast<OpT*>(tunable_op_.get());
+  }
+#endif
+
+  float alpha_;
+  float beta_;
+  bool transA_;
+  bool transB_;
+  int64_t dtype_;
+
+  // fully type erased
+  mutable std::shared_ptr<void> tunable_op_;
+};
+
+Status GemmFloat8::ComputeInternal(OpKernelContext* ctx) const {
+#if defined(DISABLE_FLOAT8_TYPES)
+  return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "DISABLE_FLOAT8_TYPES");
+#else
+  const Tensor* A = ctx->Input<Tensor>(0);
+  const Tensor* B = ctx->Input<Tensor>(1);
+  const Tensor* C = ctx->Input<Tensor>(2);  // bias
+  const Tensor* scale_a = ctx->Input<Tensor>(3);
+  const Tensor* scale_b = ctx->Input<Tensor>(4);
+  const Tensor* scale_y = ctx->Input<Tensor>(5);
+
+  auto a_shape = A->Shape();
+  auto b_shape = B->Shape();
+  ORT_ENFORCE(a_shape.NumDimensions() >= 2 && b_shape.NumDimensions() == 2);  // is in form of input @ weight
+  ORT_ENFORCE(a_shape[a_shape.NumDimensions() - 1] == b_shape[0]);            // k is compatiable
+
+  TensorShapeVector output_shape = a_shape.AsShapeVector();
+  output_shape[output_shape.size() - 1] = b_shape[b_shape.NumDimensions() - 1];
+  Tensor* Y = ctx->Output(0, output_shape);
+
+  ORT_ENFORCE(!transA_, "ROCm GemmFloat8 does not support input A transpose");
+  ORT_ENFORCE(dtype_ == onnx::TensorProto_DataType_FLOAT16, "ROCm GemmFloat8 only supports output float16");
+  ORT_ENFORCE(C == nullptr, "ROCm GemmFloat8 does not support bias input");
+  ORT_ENFORCE(scale_y == nullptr, "ROCm GemmFloat8 does not support output scaling");
+
+  if (A->IsDataType<Float8E4M3FN>()) {
+    return ComputeFp8Fp16Fp16<Float8E4M3FN>(ctx, A, scale_a, B, Y);
+  } else if (A->IsDataType<Float8E4M3FNUZ>()) {
```

Comment:
Both are emulated, native should be Float8E4M3FNUZ (in the future gen of MI300), but on MI200 it is too slow to support the full precision, so I implemented fast cvt to fp16 by flushing denorms to zeros as in trition. Since it is emulated, it is not limited to the hardware, current CDNA2 arch supports neither.